pkg/antctl/raw/multicluster/get/clusterset.go

```diff
@@ -0,0 +1,200 @@
+$ antctl mc get clusterset <clusterSetID>
+`, "\n")
+
+func (o *clusterSetOptions) validateAndComplete() {
+	if o.allNamespaces {
+		o.namespace = metav1.NamespaceAll
+		return
+	}
+	if o.namespace == "" {
+		o.namespace = metav1.NamespaceDefault
+		return
+	}
+}
+
+func init() {
+	cmdClusterSet = &cobra.Command{
+		Use: "clusterset",
+		Aliases: []string{
+			"clustersets",
+		},
+		Short:   "Print Multi-cluster ClusterSets",
+		Args:    cobra.MaximumNArgs(1),
+		Example: clusterSetExamples,
+		RunE:    runEClusterSet,
+	}
+	o := &clusterSetOptions{}
+	optionsClusterSet = o
+	cmdClusterSet.Flags().StringVarP(&o.namespace, "namespace", "n", "", "Namespace of ClusterSets")
+	cmdClusterSet.Flags().StringVarP(&o.outputFormat, "output", "o", "", "Output format. Supported formats: json|yaml")
+	cmdClusterSet.Flags().BoolVarP(&o.allNamespaces, "all-namespaces", "A", false, "If present, list ClusterSets across all namespaces")
+
+	multicluster.GetCmd.AddCommand(cmdClusterSet)
+}
+
+func runEClusterSet(cmd *cobra.Command, args []string) error {
+	optionsClusterSet.validateAndComplete()
+
+	kubeconfig, err := raw.ResolveKubeconfig(cmd)
+	if err != nil {
+		return err
+	}
+
+	argsNum := len(args)
+	singleResource := false
+	if argsNum > 0 {
+		singleResource = true
+	}
+
+	k8sClient, err := raw.AddSchemeToClientCfg(kubeconfig)
+	if err != nil {
+		return err
+	}
+
+	var clusterSets []multiclusterv1alpha1.ClusterSet
+	if singleResource {
+		clusterSetName := args[0]
+		clusterSet := multiclusterv1alpha1.ClusterSet{}
+		err = k8sClient.Get(context.TODO(), types.NamespacedName{
+			Namespace: optionsClusterSet.namespace,
+			Name:      clusterSetName,
+		}, &clusterSet)
+		if err != nil {
+			return err
+		}
+		gvks, unversioned, err := k8sClient.Scheme().ObjectKinds(&clusterSet)
+		if err != nil {
+			return err
+		}
+		if !unversioned && len(gvks) == 1 {
+			clusterSet.SetGroupVersionKind(gvks[0])
+		}
+		clusterSets = append(clusterSets, clusterSet)
+	} else {
+		clusterSetList := &multiclusterv1alpha1.ClusterSetList{}
+		err = k8sClient.List(context.TODO(), clusterSetList, &client.ListOptions{Namespace: optionsClusterSet.namespace})
+		if err != nil {
+			return err
+		}
+		clusterSets = clusterSetList.Items
+	}
+
+	if len(clusterSets) == 0 {
+		if optionsClusterSet.namespace != "" {
+			fmt.Printf("No resource found in Namespace %s\n", optionsClusterSet.namespace)
+		} else {
+			fmt.Println("No resource found in all Namespaces")
+		}
+		return nil
+	}
+
+	switch optionsClusterSet.outputFormat {
+	case "json":
+		err := antctl.JsonOutput(clusterSets, os.Stdout)
+		if err != nil {
+			return err
+		}
+	case "yaml":
+		err := antctl.YamlOutput(clusterSets, os.Stdout)
+		if err != nil {
+			return err
+		}
+	default:
+		if singleResource {
+			fmt.Println(outputClusterSet(clusterSets))
+		} else {
+			fmt.Println(outputAllClusterSet(clusterSets))
+		}
+	}
+
+	return nil
+}
+
+func outputClusterSet(clusterSets []multiclusterv1alpha1.ClusterSet) string {
```

Comment:
Can we implement TableOutput and use output() as other commands?

Reply:
@jianjuns  I tried this idea, I think technically this makes sense for me, only one format problem here.

`CLUSTERID NAMESPACE CLUSTERSETID    TYPE            STATUS REASON
cluster-a leader-ns test-clusterset IsElectedLeader True   <NONE>
cluster-a leader-ns test-clusterset Ready           True   <NONE>

CLUSTERID  NAMESPACE                 CLUSTERSETID     TYPE            STATUS REASON
cluster-a1 leader-nsforholdingalotof test-clusterset1 IsElectedLeader True   <NONE>
cluster-a1 leader-nsforholdingalotof test-clusterset1 Ready           True   <NONE>
`

Because of the format we use, we cannot align the columns for every output(). In this case, when the lengths of namespaces are too different between ClusterSet, it will cause visual confusion: under the first Type is the second ClusterSetID, so if we still want this, is it good to change the format for all output()? Or we use a specific format for ClusterSet.